adapter/internal/ga/client.go

```diff
@@ -0,0 +1,214 @@
+/*
+ *  Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package ga
+
+import (
+	"context"
+	"io"
+
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/wso2/product-microgateway/adapter/config"
+	logger "github.com/wso2/product-microgateway/adapter/internal/loggers"
+	ga_model "github.com/wso2/product-microgateway/adapter/pkg/discovery/api/wso2/discovery/ga"
+	stub "github.com/wso2/product-microgateway/adapter/pkg/discovery/api/wso2/discovery/service/ga"
+	"google.golang.org/genproto/googleapis/rpc/status"
+	"google.golang.org/grpc"
+)
+
+var (
+	// apiRevision Map keeps apiUUID -> revisionUUID. This is used only for the communication between global adapter and adapter
+	// The purpose here is to identify if the certain API's revision is already added to the XDS cache.
+	apiRevisionMap map[string]string
+	// Last Acknowledged Response from the global adapter
+	laskAckedResponse *discovery.DiscoveryResponse
+	// Last Received Response from the global adapter
+	// Last Recieved Response is always is equal to the lastAckedResponse according to current implementation as there is no
+	// validation performed on successfully recieved response.
+	lastReceivedResponse *discovery.DiscoveryResponse
+	// XDS stream for streaming APIs from Global Adapter
+	xdsStream stub.ApiGADiscoveryService_StreamGAApisClient
+	// GAAPIChannel stores the API Events composed from XDS states
+	GAAPIChannel chan *APIEvent
+)
+
+const (
+	// The type url for requesting API Entries from global adapter.
+	apiTypeURL string = "type.googleapis.com/wso2.discovery.ga.Api"
+)
+
+// APIEvent represents the event corresponding to a single API Deploy or Remove event
+// based on XDS state changes
+type APIEvent struct {
+	APIUUID       string
+	RevisionUUID  string
+	IsDeployEvent bool
+}
+
+func init() {
+	apiRevisionMap = make(map[string]string)
+	laskAckedResponse = &discovery.DiscoveryResponse{}
+	GAAPIChannel = make(chan *APIEvent)
+}
+
+func initConnection(xdsURL string) error {
+	// TODO: (VirajSalaka) Close Connection
```

Comment:
What is the connection close condition?

Reply:
removed via https://github.com/wso2/product-microgateway/pull/2107/commits/13c129233dc5f1ab801d14f9d923d1a6bc2d8752
closing the connection would be required when retrying is implemented.